pkg/webhook/resourcedistribution/validating/resourcedistribution_create_update_handler.go

```diff
@@ -0,0 +1,202 @@
+func (h *ResourceDistributionCreateUpdateHandler) validateResourceDistributionResourceKindAndSemantics(resource runtime.Object, fldPath *field.Path) (allErrs field.ErrorList) {
+	// 1. check whether the GVK of the resource is in supportedGVKList
+	if !isSupportedGVK(resource) {
+		return append(allErrs, field.Invalid(fldPath, resource, fmt.Sprintf("unknown or unsupported resource GroupVersionKind, only support %v", supportedGVKList)))
+	}
+	// 2. dry run to check resource
+	mice := resource.DeepCopyObject()
+	ConvertToUnstructured(mice).SetNamespace(DefaultNamespace)
+	if err := h.Client.Create(context.TODO(), mice, &client.CreateOptions{
+		DryRun: []string{metav1.DryRunAll},
+	}); err != nil {
+		return append(allErrs, field.InternalError(fldPath, fmt.Errorf("create namespaces failed, err: %v", err)))
+	}
+	return
+}
+
+// validateResourceDistributionSpecTargets validate Spec.Targets
+// (1). parse target namespaces
+// (2). validate conflict between existing resources
+func (h *ResourceDistributionCreateUpdateHandler) validateResourceDistributionSpecTargets(targets *appsv1alpha1.ResourceDistributionTargets, fldPath *field.Path) (allErrs field.ErrorList) {
+	// 1. validate namespace
+	forbidden := make([]string, 0)
+	conflicted := make([]string, 0)
+	included := make(map[string]struct{})
+	for _, namespace := range targets.IncludedNamespaces.List {
+		included[namespace.Name] = struct{}{}
+		// validate namespace name
+		for _, msg := range coreval.ValidateNamespaceName(namespace.Name, false) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("includedNamespaces"), targets.IncludedNamespaces, msg))
+		}
+		// validate whether namespace is forbidden
+		if isForbiddenNamespace(namespace.Name) {
+			forbidden = append(forbidden, namespace.Name)
+		}
+	}
+	for _, namespace := range targets.ExcludedNamespaces.List {
+		// validate namespace name
+		for _, msg := range coreval.ValidateNamespaceName(namespace.Name, false) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("excludedNamespaces"), targets.ExcludedNamespaces, msg))
+		}
+		// validate conflict between IncludedNamespaces and ExcludedNamespaces
+		if _, ok := included[namespace.Name]; ok {
+			conflicted = append(conflicted, namespace.Name)
+		}
+	}
+	if len(conflicted) != 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath, targets, fmt.Sprintf("ambiguous targets because namespace %v is in both IncludedNamespaces.List and ExcludedNamesapces.List", conflicted)))
+	}
+	if len(forbidden) != 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("includedNamespaces"), targets.IncludedNamespaces, fmt.Sprintf("cannot distribute rsource to forbidden namespaces %v", ForbiddenNamespaces)))
+	}
+
+	// 2. validate targets.NamespaceLabelSelector
+	allErrs = append(allErrs, metav1validation.ValidateLabelSelector(&targets.NamespaceLabelSelector, fldPath.Child("namespaceLabelSelector"))...)
```

Comment:
it is better to initiate the selector by calling LabelSelectorAsSelector, LabelSelectorAsSelector can validate the label more strictly than ValidateLabelSelector

Reply:
done